### PR TITLE
Fix building empty MLIR tuples

### PR DESCRIPTION
**Pull request by jonatanklosko**
Currently we infer `%Function{}` from tuple arguments, but if we want to build empty tuple, there are none. And we even build such explicitly here:

https://github.com/elixir-nx/nx/blob/c5d91781b91f63cae4917e75e0dc77507852e608/exla/lib/exla/defn.ex#L721

### CODE REVIEW

File: exla/lib/exla/mlir/value.ex [hunk 1]
```diff
@@ -76,8 +76,8 @@ defmodule EXLA.MLIR.Value do
     %Value{op | ref: ref}
   end
 
-  def tuple([%Value{function: %Function{} = func} | _rest] = vals) do
-    refs = Enum.map(vals, fn %Value{ref: ref, function: ^func} -> ref end)
```
Comment (jonatanklosko) [inline exla/lib/exla/mlir/value.ex:80]: We no longer assert on the same function, but we don't that for the `dynamic_slice` above either. @polvalente is this fine?

**Review by josevalim — APPROVED** [review]

File: exla/lib/exla/mlir/value.ex (hunk 1 above)
Comment (josevalim) [reply exla/lib/exla/mlir/value.ex:80]: Yeah, I think it is fine. I think it would be very hard in practice to mix values from different functions.

Comment (jonatanklosko) [reply exla/lib/exla/mlir/value.ex:80]: There was a case where the builder function was different than the vals functions, that's why I removed the assertion, but that's a different case?

Comment (josevalim) [reply exla/lib/exla/mlir/value.ex:80]: Hrm.... we need to wait for @polvalente but maybe this clause was never actually invoked and appeared by accident?

Comment (polvalente) [reply exla/lib/exla/mlir/value.ex:80]: @jonatanklosko I checked out on this branch and put the `Enum.map` assertion back and all tests still passed. I think we can restore it.

Every time I faced a similar problem, the actual fix was elsewhere.

This tends to happen when building cond and we end up using a value from the root scope inside the if-else block (each is built as a separate function).

**Review by polvalente — APPROVED** [review]
LGTM, we just need to decide on whether to restore the function assertion

File: exla/lib/exla/mlir/value.ex (hunk 1 above)
Comment (jonatanklosko) [reply exla/lib/exla/mlir/value.ex:80]: It does fail some tests in bumblebee though. Here is a minimal reproduction:

<details><summary>.livemd</summary>

````markdown
# mlir tuple

```elixir
Mix.install([
  {:nx, path: "~/git/nx/nx", override: true},
  {:exla, path: "~/git/nx/exla", override: true}
])

Nx.global_default_backend(EXLA.Backend)
Nx.Defn.global_default_options(compiler: EXLA, compiler_mode: :mlir)
```

## minimized

```elixir
defmodule Test do
  import Nx.Defn

  defn test(tensor, condition) do
    if condition do
      tensor
    else
      {_i, tensor} =
        while {i = 0, tensor}, i < 1 do
          {i + 1, tensor}
        end

      tensor
    end
  end
end
```

```elixir
tensor = Nx.tensor([1.0, 2.0, 3.0, 4.0])
condition = Nx.tensor(true)
Test.test(tensor, condition)
```
````

</details>

Comment (polvalente) [reply exla/lib/exla/mlir/value.ex:80]: Yeah, with this test I agree with removing the assertion. Basically what happens is that we build each branch of the cond as a separate function, and then one of the while tuples ends up trying to create a tuple with the root-scoped `tensor` and the else-scoped `i`, which then fails the assertion.

But these extraneous ocurrences are rewritten when we're splicing the cond together (same with while), and then everything ends up ok.

Comment (jonatanklosko) [reply exla/lib/exla/mlir/value.ex:80]: Makes sense!

**Pull request merged**